core/src/main/java/org/apache/iceberg/DefaultValueParser.java

```diff
@@ -0,0 +1,253 @@
+        return jsonNode.booleanValue();
+      case INTEGER:
+        return jsonNode.intValue();
+      case LONG:
+        return jsonNode.longValue();
+      case FLOAT:
+        return jsonNode.floatValue();
+      case DOUBLE:
+        return jsonNode.doubleValue();
+      case DECIMAL:
+        return jsonNode.decimalValue();
+      case STRING:
+      case UUID:
+        return jsonNode.textValue();
+      case DATE:
+      case TIME:
+      case TIMESTAMP:
+        return Literal.of(jsonNode.textValue()).to(type).value();
+      case FIXED:
+        byte[] fixedBytes = BaseEncoding.base16().decode(jsonNode.textValue().toUpperCase(Locale.ROOT).replaceFirst(
+            "^0X",
+            ""));
+        return ByteBuffer.allocate(((Types.FixedType) type).length()).put(fixedBytes);
+      case BINARY:
+        byte[] binaryBytes = BaseEncoding.base16().decode(jsonNode.textValue().toUpperCase(Locale.ROOT).replaceFirst(
+            "^0X", ""));
+        return ByteBuffer.wrap(binaryBytes);
+      case LIST:
+        List<Object> defaultList = Lists.newArrayList();
+        for (JsonNode element : jsonNode) {
+          defaultList.add(parseDefaultFromJson(type.asListType().elementType(), element));
+        }
+        return defaultList;
+      case MAP:
+        Map<Object, Object> defaultMap = Maps.newHashMap();
+        List<JsonNode> keysAndValues = StreamSupport
+            .stream(jsonNode.spliterator(), false)
+            .collect(Collectors.toList());
+        JsonNode keys = keysAndValues.get(0);
+        JsonNode values = keysAndValues.get(1);
+
+        List<JsonNode> keyList = Lists.newArrayList(keys.iterator());
+        List<JsonNode> valueList = Lists.newArrayList(values.iterator());
+
+        for (int i = 0; i < keyList.size(); i++) {
+          defaultMap.put(
+              parseDefaultFromJson(type.asMapType().keyType(), keyList.get(i)),
+              parseDefaultFromJson(type.asMapType().valueType(), valueList.get(i)));
+        }
+        return defaultMap;
+      case STRUCT:
+        Map<Integer, Object> defaultStruct = Maps.newHashMap();
+        for (Types.NestedField subField : type.asStructType().fields()) {
+          String fieldIdAsString = String.valueOf(subField.fieldId());
+          Object value = jsonNode.has(fieldIdAsString) ? parseDefaultFromJson(
+              subField.type(),
+              jsonNode.get(fieldIdAsString)) : null;
+          if (value != null) {
+            defaultStruct.put(subField.fieldId(), value);
+          }
+        }
+        return defaultStruct;
+      default:
+        return null;
+    }
+  }
+
+  public static JsonNode validateDefault(String name, Type type, JsonNode defaultValue) {
+    if (defaultValue != null && !isValidDefault(type, defaultValue)) {
+      throw new ValidationException("Invalid default value for field %s: %s not a %s", name, defaultValue, type);
+    }
+    return defaultValue;
+  }
+
+  @SuppressWarnings("checkstyle:CyclomaticComplexity")
```

Comment:
We generally don't suppress this warning unless we are trying to avoid major code changes. Since this is all new, it is best to follow the recommendation.